File: grid.py
from math import floor
from typing import List
from particles import Particle
from config import SCREEN_WIDTH, SCREEN_HEIGHT
from particles import PARTICLE_WIDTH, PARTICLE_HEIGHT, Sand

class Grid:
    _rows: int # 48
    _cols: int # 64
    _grid: List[List[Particle]]

    def __init__(self):
        self._grid = self._create_empty_grid()

    @property
    def rows(self):
        return self._rows

    @property
    def cols(self):
        return self._cols


    def _create_empty_grid(self) -> List[List[None]]:
        self._rows, self._cols = round(SCREEN_HEIGHT / PARTICLE_HEIGHT), round(SCREEN_WIDTH / PARTICLE_WIDTH)
        grid = [[None for _ in range(self.cols)] for _ in range(self.rows)]
        print(self.rows, self.cols)
        return grid

    def traverse_grid(self, window) -> None:
        for row in reversed(range(self.rows)):
            for col in reversed(range(self.cols)):
                particle = self._grid[row][col]
                if particle:
                #     particle.update(self._grid)
                    particle.draw(window, col, row)

    def create_particle(self, mouse_position: tuple[int, int]) -> None:
        x, y = mouse_position
        if not self._mouse_within_screen(x, y):
            return

        col, row = floor(x/10), floor(y/10)
        if self._grid[row][col] is None:
            self._grid[row][col] = Sand()


    def _mouse_within_screen(self, x: int, y: int) -> bool:
        return 0 <= x <= SCREEN_WIDTH - PARTICLE_WIDTH and 0 <= y <= SCREEN_HEIGHT - PARTICLE_HEIGHT

File: src/particles/Grid.java
package particles;

public class Grid {
    private final int rows;
    private final int cols;
    private final Particle[][] grid;

    public Grid(int rows, int cols) {
        this.rows = rows;
        this.cols = cols;
        this.grid = new Particle[rows][cols];
    }

    public void updateGrid() {
        for (int row = rows-1; row >= 0; row--) {
            for (int col = 0; col < cols; col++) {
                Particle p = grid[row][col];
                if (p != null)
                    p.update(row, col, this);
            }
        }
    }

    public void setParticle(int row, int col, Particle particle) {
        grid[row][col] = particle;
    }

    public Particle getParticle(int row, int col) {
        return grid[row][col];
    }

    public int getRows() {
        return rows;
    }

    public int getCols() {
        return cols;
    }

    public Particle[][] getGrid() {
        return grid;
    }
}
